r1cs/core/src/constraint_system.rs

```diff
@@ -73,6 +93,21 @@ pub trait ConstraintSystemAbstract<F: Field>: Sized {
 
     /// Output the number of constraints in the system.
     fn num_constraints(&self) -> usize;
+
+    /// Start/restart counting the constraints associated to a given label. This function allows to
+    /// specify that all the constraints enforced after calling this function should be counted in
+    /// a specific counter identified by `counter_label`.
+    fn restart_constraints_counter(&mut self, counter_label: &str);
+
+    /// Stop counting the constraints associated to a given label. This function undoes a previous
+    /// call to `restart_constraints_counter`, specifying that all the constraints enforced after
+    /// calling this function should no longer be counted in the specific counter identified by
+    /// `counter_label`. The counting can be resumed by calling `restart_constraints_counter`
+    /// passing the same label
+    fn stop_constraints_counter(&mut self, counter_label: &str) -> Result<(), SynthesisError>;
+
+    /// Get the current value of the counter for constraints associated to `counter_label`
+    fn get_constraints_counter(&mut self, counter_label: &str) -> Result<usize, SynthesisError>;
```

Comment:
I don't like these functions being in the ConstraintSystemAbstract trait, It's not possible to add them to the ConstraintSystemDebugger trait right ? Otherwise you need to add its trait bound to each CS: ConstraintSystemAbstract template.
Perhaps we can do it by making ConstraintSystemDebugger an associated type of the ConstraintSystemAbstract trait ? Seems quite legit.
Anyway let's not waste time on this now, and let's put a TODO instead

Reply:
Agree that it is not so nice to have it in `ConstraintSystemAbstract`, but it was necessary for the reason you mentioned. I will add a TODO to be addressed in case of refactoring of `ConstraintSystemAbstract`.